### PR TITLE
Cache responses from the A+ API

### PR DESCRIPTION
**Pull request by nikke234**
# Description of the PR

This adds caching for A+ responses. It also uses the caching by downloading all submission results now from the API end point of an individual submission, which fixes the issue of incorrect statuses in submissions. The cache file can get relatively large, so there's some logic for doing IO in a background thread and combining multiple file writes when there's a burst of insertions into the cache.

~Beware in testing:
Loading the exercises initially will take some time, since it's doing quite many requests when no cache exists yet. There is no indication of progress, so it might appear as if the exercises are simply broken. As such, I would also consider this blocked by #609.~
Fixed by adding proxy objects, thanks to @OlliKiljunen for the idea

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [x] new <b>unit</b> tests created</li>
        <li>- [x] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [x] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [x] Yes.
- [x] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

**Comment by OlliKiljunen** [comment]
Do you think it would be possible to show the assignment tree first with "stupid" [proxy](https://en.wikipedia.org/wiki/Proxy_pattern) objects constructed from the data got from the "get all exercises" API call, and then "realize" them with the actual data when it's there?

**Comment by nikke234** [comment]
> Do you think it would be possible to show the assignment tree first with "stupid" [proxy](https://en.wikipedia.org/wiki/Proxy_pattern) objects constructed from the data got from the "get all exercises" API call, and then "realize" them with the actual data when it's there?

Good idea! Will do

**Comment by OlliKiljunen** [comment]
Just a minor issue, but crossed my mind. There's **.gitignore** file in each project's **.idea** folder. Should **a-plus-cache.json** be added there?

**Comment by nikke234** [comment]
> Just a minor issue, but crossed my mind. There's **.gitignore** file in each project's **.idea** folder. Should **a-plus-cache.json** be added there?

Yeah that would be good. It will work fine when switching machines, but it will generate some pointless diffs though. Is there some IntelliJ API for that or shall I just append a line to it?

**Comment by OlliKiljunen** [comment]
> 
> 
> > Just a minor issue, but crossed my mind. There's **.gitignore** file in each project's **.idea** folder. Should **a-plus-cache.json** be added there?
> 
> Yeah that would be good. It will work fine when switching machines, but it will generate some pointless diffs though. Is there some IntelliJ API for that or shall I just append a line to it?

I don't know, but...

Actually, could or should that file being stored somewhere else like where IntelliJ itself stores caches/settings? For example, it seems there's a folder named **sonarlint** in **%APPDATA%\JetBrains\IdeaIC2021.1** folder (so called [system directory](https://www.jetbrains.com/help/idea/tuning-the-ide.html#system-directory) of IntelliJ IDEA), and there are data used by SonarLint plugin.

Even though cached server responses are kind of related to a certain project, on the other hand, they are not, if you got what I mean.

**Comment by nikke234** [comment]
> > > Just a minor issue, but crossed my mind. There's **.gitignore** file in each project's **.idea** folder. Should **a-plus-cache.json** be added there?
> > 
> > 
> > Yeah that would be good. It will work fine when switching machines, but it will generate some pointless diffs though. Is there some IntelliJ API for that or shall I just append a line to it?
> 
> I don't know, but...
> 
> Actually, could or should that file being stored somewhere else like where IntelliJ itself stores caches/settings? For example, it seems there's a folder named **sonarlint** in **%APPDATA%\JetBrains\IdeaIC2021.1** folder (so called [system directory](https://www.jetbrains.com/help/idea/tuning-the-ide.html#system-directory) of IntelliJ IDEA), and there are data used by SonarLint plugin.
> 
> Even though cached server responses are kind of related to a certain project, on the other hand, they are not, if you got what I mean.

That's definitely a possibility assuming there are no permission issues with writing there (which there shouldn't be).

It's true that it's kinda not project related. After all, the same URL should yield the same result regardless of which course a project is using. If that approach is taken, should the code also then have a singleton class available from every project to reflect that?

**Comment by OlliKiljunen** [comment]
> 
> 
> > > > Just a minor issue, but crossed my mind. There's **.gitignore** file in each project's **.idea** folder. Should **a-plus-cache.json** be added there?
> > > 
> > > 
> > > Yeah that would be good. It will work fine when switching machines, but it will generate some pointless diffs though. Is there some IntelliJ API for that or shall I just append a line to it?
> > 
> > 
> > I don't know, but...
> > Actually, could or should that file being stored somewhere else like where IntelliJ itself stores caches/settings? For example, it seems there's a folder named **sonarlint** in **%APPDATA%\JetBrains\IdeaIC2021.1** folder (so called [system directory](https://www.jetbrains.com/help/idea/tuning-the-ide.html#system-directory) of IntelliJ IDEA), and there are data used by SonarLint plugin.
> > Even though cached server responses are kind of related to a certain project, on the other hand, they are not, if you got what I mean.
> 
> That's definitely a possibility assuming there are no permission issues with writing there (which there shouldn't be).
> 
> It's true that it's kinda not project related. After all, the same URL should yield the same result regardless of which course a project is using. If that approach is taken, should the code also then have a singleton class available from every project to reflect that?

Possibly... I'll consider that when reviewing your code. 👍

**Review by OlliKiljunen — APPROVED** [review]
Wow, a very nice implementation.

I can think of at least three possible ways to cache:
- A) what is done here: cache kept in memory; every so often synced to a file
- B) similar to what is done here, but cache is not kept in memory between the requests, or it is swiped from the memory after a while in inactivity
- C) there's no in-memory cache but cache entries are stored as (separate) files

From A to C, the memory usage decreases, on the other hand, there become more file system accesses and time efficiency is reduced.

I think your choice is justifiable. I estimate the memory footprint of the cache to be approx. ~1MB, which is not necessarily too much considering the totality of IntelliJ being easily ~1GB even in idle.

Particularly, I like how your design allows us to change between options A, B and C if ever needed. (If we wanted to, we could extract interfaces of `Cache` and `Entry`'s public methods to make sure the client code doesn't become dependent on a certain implementation.)

Btw, very nice threading. 🧵 

Approved! 👍

File: src/main/java/fi/aalto/cs/apluscourses/dal/APlusExerciseDataSource.java [hunk 1]
```diff
@@ -149,10 +154,11 @@ public Points getPoints(@NotNull Course course, @NotNull Authentication authenti
...
   public SubmissionResult getSubmissionResult(@NotNull String submissionUrl,
                                               @NotNull Exercise exercise,
-                                              @NotNull Authentication authentication)
+                                              @NotNull Authentication authentication,
+                                              @NotNull ZonedDateTime minCacheEntryTime)
       throws IOException {
-    JSONObject response = client.fetch(submissionUrl, authentication);
-    return parser.parseSubmissionResult(response, exercise);
+    JSONObject response = client.fetch(submissionUrl, authentication, minCacheEntryTime);
+    return SubmissionResult.fromJsonObject(response, exercise);
```
[the remark this answers is not in the document]
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/dal/APlusExerciseDataSource.java:161]: Oops, testing left-over

File: src/main/java/fi/aalto/cs/apluscourses/model/Cache.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+ * A cache is essentially a map backed by a file. For each value, the time of creation is also
+ * stored so that clients can determine the age of the cache entry. A subclass needs to implement
+ * {@link Cache#fromFile} and {@link Cache#toFile}, which read and write the entries to a file. The
+ * file format is completely up to the client. The file is read once when the cache is instantiated.
+ * Writes occur after each insertion, but bursts of insertions only cause one write. Reading and
+ * writing is done in a background thread, so a read or write may take a long time without blocking
+ * other code. Note, that the cache makes no guarantee that every entry inserted is also written to
+ * the file.
+ */
+public abstract class Cache<KeyT, ValueT> {
```
[the remark this answers is not in the document]
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/model/Cache.java:27]: Sonar on the other hand does complain even with the T, so perhaps I'll go with K and V after all :frowning_face:

**Review by stellatsv — APPROVED** [review]
Excellent work!

**Pull request merged**